compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEInsertAIEWorkgroup.cpp

```diff
@@ -0,0 +1,169 @@
+    return WalkResult::advance();
+  });
+  return success();
+}
+
+/// Insert core ops inside innermost forall ops around computational ops and
+/// add synchronization ops along the way to synchronize with surrounding
+/// dma ops.
+LogicalResult insertCoreOpsInWorkgroup(mlir::ModuleOp moduleOp) {
+  IRRewriter rewriter(moduleOp.getContext());
+  WalkResult res = moduleOp->walk([&](AMDAIE::WorkgroupOp workgroupOp) {
+    // Find innermost forall loops
+    SmallVector<scf::ForallOp> innermostForallLoops =
+        getInnermostForallLoops(workgroupOp);
+    if (innermostForallLoops.empty()) {
+      workgroupOp.emitOpError() << "contains no nested forall op";
+      return WalkResult::interrupt();
+    }
+    if (innermostForallLoops.size() != 1) {
+      workgroupOp.emitOpError()
+          << "multiple innermost foralls not supported for now";
+      return WalkResult::interrupt();
+    }
+    scf::ForallOp innermostForall = innermostForallLoops[0];
+    auto parentOps = getParentsOfTypeIncluding<scf::ForallOp>(innermostForall);
+    DenseMap<Attribute, Value> attrMapping;
+    getAttributeMapping(parentOps, attrMapping);
+    if (!attrMapping.contains(gpu::threadX(workgroupOp->getContext())) ||
+        !attrMapping.contains(gpu::threadY(workgroupOp->getContext()))) {
+      workgroupOp.emitOpError() << "no forall with thread mapping found";
+      return WalkResult::interrupt();
+    }
+    Value threadX = attrMapping[gpu::threadX(workgroupOp->getContext())];
+    Value threadY = attrMapping[gpu::threadY(workgroupOp->getContext())];
+    // Create CoreOp at the end of the innermost forall
+    rewriter.setInsertionPoint(innermostForall.getBody()->getTerminator());
+    auto coreOp = rewriter.create<AMDAIE::CoreOp>(rewriter.getUnknownLoc(),
+                                                  threadX, threadY);
+    Region &region = coreOp.getRegion();
+    Block *newBlock = rewriter.createBlock(&region);
+    rewriter.setInsertionPointToStart(newBlock);
+    auto endOp = rewriter.create<AMDAIE::EndOp>(rewriter.getUnknownLoc());
+
+    // Walk all operations in the workgroup and fill in the CoreOp with
+    // computational ops (linalg) and synchronization ops to synchronize
+    // with the workgroup DMA ops.
+    WalkResult workgroupRes = workgroupOp->walk([&](Operation *op) {
+      // Skip operations already inside core ops
+      if (op->getParentOfType<AMDAIE::CoreOp>()) return WalkResult::advance();
+      if (auto dmaOp = dyn_cast<AMDAIE::DmaCpyNdOp>(op)) {
+        auto sourceMemspace = dmaOp.getSourceObjectFifo().getMemorySpace();
+        auto targetMemspace = dmaOp.getTargetObjectFifo().getMemorySpace();
+        if (sourceMemspace &&
+            dyn_cast<IntegerAttr>(sourceMemspace).getInt() == 2 &&
+            targetMemspace &&
+            dyn_cast<IntegerAttr>(targetMemspace).getInt() == 2) {
+          dmaOp->emitOpError()
+              << "dma op with both source and target on L2 is not supported";
```

Comment:
Why do you just check both source and target on L2 but not L1?

Reply:
I think I made a type here, should be `L1`. But then I guess you would have the same question for `L2`? It's not checked for because both source and target on L2 will not enter any `if/else` block and will just be skipped. The logic here is just concerned about DMA operations taking data to/from L1 to insert the necessary synchronization ops.